### PR TITLE
fix: Specifying @link in config skips upstream.allowedHeaders

### LINKED ISSUE #1683: bug(config): Specifying @link in config breaks upstream.allowedHeaders setting
### Prerequisites

- [x] I have checked the existing issues to ensure this is not a duplicate.
- [x] I am running the latest version of Tailcall.

### Describe the bug

Take this config

```graphql
schema
  @server(port: 8000, graphiql: true)
  @link(src: "./stripe-types.graphql", type: Config)
  @upstream(baseURL: "https://api.stripe.com/v1/", allowedHeaders: ["Authorization"]) {
  query: Query
}

type Query {
  # BALANCE
  balance: BalanceRoot
    @http(
      method: "GET"
      path: "/balance"
    )
}

```

and 

```graphql
# Balance Schema

type Balance {
  amount: Int
  currency: String
  source_types: SourceTypes
}

type SourceTypes {
  card: Int
}

type ConnectReserved {
  amount: Int
  currency: String
}

type BalanceRoot {
  object: String
  livemode: Boolean
  pending: [Balance]
  connect_reserved: [ConnectReserved]
  available: [Balance]
}
```

If we run this example and try to pass the Authorization header we'll get an auth error from the upstream. 

And if we combine these files into single config file everything will work just fine.

### Steps to reproduce

1. Use config from above
2. Start the tailcall
3. Try to pass authorization header

### Expected behavior

Authorization header should be passed to the upstream if it's defined in user's request

### Actual behavior

The Authorization header is ignored

### Additional context

The bug happens because of the implementation of `merge_right` for Config::Upstream here https://github.com/tailcallhq/tailcall/blob/b86b37bb4471ff07db24c63d84e7c59df6e3a49e/src/config/upstream.rs#L173-L180 - in this code in case `other.allowedHeaders` is None then `self.allowedHeaders` is ignored completely and the result would be None.

### PR DESCRIPTION
**Pull request by amitksingh1490**
fixes #1683

### CODE REVIEW

File: src/merge_right.rs [hunk 1]
```diff
@@ -1,13 +1,61 @@
...
+
+impl Scalar for char {}
+
+impl Scalar for String {}
+
+impl Scalar for HttpVersion {}
+
+impl Scalar for ScriptOptions {}
+
+impl Scalar for Proxy {}
```
[the remark this answers is not in the document]
Comment (amitksingh1490) [reply src/merge_right.rs:54]: Why is this a problem

File: src/merge_right.rs [hunk 2]
```diff
@@ -1,13 +1,61 @@
...
+
+impl Scalar for bool {}
+
+impl Scalar for char {}
+
+impl Scalar for String {}
+
+impl Scalar for HttpVersion {}
+
+impl Scalar for ScriptOptions {}
```
Comment (tusharmath) [inline src/merge_right.rs:52]: I don't think ScriptOptions is a a scalar.

Comment (amitksingh1490) [reply src/merge_right.rs]: done

File: src/merge_right.rs [hunk 3]
```diff
@@ -1,13 +1,65 @@
...
+
+impl Scalar for bool {}
+
+impl Scalar for char {}
+
+impl Scalar for String {}
+
+impl Scalar for HttpVersion {}
+
+impl MergeRight for ScriptOptions {
```
Comment (tusharmath) [inline src/merge_right.rs:52]: Move the implementation closer to where the type is located.

File: src/merge_right.rs [hunk 4]
```diff
@@ -1,13 +1,65 @@
...
+    fn merge_right(self, other: Self) -> Self {
+        ScriptOptions { timeout: self.timeout.merge_right(other.timeout) }
+    }
+}
+
+impl Scalar for Proxy {}
```
Comment (tusharmath) [inline src/merge_right.rs:58]: Proxy isn't a Scalar.

File: src/merge_right.rs [hunk 5]
```diff
@@ -1,13 +1,65 @@
...
+
+impl<A: Scalar> MergeRight for A {
```
Comment (tusharmath) [inline src/merge_right.rs:60]:
```suggestion
impl MergeRight for Scalar {
```

File: src/merge_right.rs [hunk 6]
```diff
@@ -1,13 +1,65 @@
...
+
+impl Scalar for f32 {}
+
+impl Scalar for bool {}
+
+impl Scalar for char {}
+
+impl Scalar for String {}
+
+impl Scalar for HttpVersion {}
```
Comment (tusharmath) [inline src/merge_right.rs:50]: Create a separate file under src called `scalar`. Where all the scalars are implemented.

Comment (tusharmath) [reply src/merge_right.rs:50]: We can even call it `Primitive` since graphQL scalars already exist.

Comment (amitksingh1490) [reply src/merge_right.rs]: done file called primitives

File: src/config/server.rs [hunk 7]
```diff
@@ -195,6 +195,7 @@ impl Server {
         self.pipeline_flush.unwrap_or(true)
     }
 }
+
```
[the remark this answers is not in the document]
Comment (amitksingh1490) [reply src/config/server.rs:210]: untouched code path, we should do it in separate PR

File: tests/execution/test-merge-right-with-link-config.md [hunk 8]
```diff
@@ -0,0 +1,45 @@
+---
+check_identity: false
+---
+
+# test-merge-right-with-link-config
```
[the remark this answers is not in the document]
Comment (amitksingh1490) [reply tests/execution/test-merge-right-with-link-config.md:5]: I want the file name to be same as test name

File: src/config/server.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (amitksingh1490) [reply src/config/server.rs:210]: sure open an issue

**Review by tusharmath — APPROVED** [review]

**Pull request merged**